standard/types.md

```diff
@@ -27,10 +27,15 @@ For convenience, throughout this specification, some library type names are writ
 
 ### 8.2.1 General
 
-A reference type is a class type, an interface type, an array type, a delegate type, or the `dynamic` type.
+A reference type is a class type, an interface type, an array type, a delegate type, or the `dynamic` type. For each non-nullable reference type, there is a corresponding nullable reference type noted by appending the `?` to the type name.
 
 ```ANTLR
 reference_type
+    : non_nullable_reference_type
+    | nullable_reference_type
+    ;
+
+non_nullable_reference_type
```

Comment:
We’re stretching for rules names, and I’m not sure of a good solution. This one isn't tortuous in the way *non_nullable_non_value_type_parameter* (#1178), rather it is somewhat misleading.

If NRT checking is disabled all these types are nullable. If it is enabled then these types are still nullable but may raise warnings if an implementation determines they are assigned null.

But what alternatives can we come up with?

**Just musing**: *basic_reference_type*? Maybe *reference_type* ⇒ *general_reference_type* allowing *non_nullable_reference_type* ⇒ *reference_type*? The latter would allow text to state that something of *reference_type* may be assigned null, which in turn may elicit a warning if NRT is enabled. That has to be better than saying something of *non_nullable_reference_type* may be assigned null surely?

Of course in some future version of the Standard we may well be describing the disabled NRT state as a little used feature for backward compatibility and then what is being written now won’t be jarring…

Reply:
> If NRT checking is disabled all these types are nullable.

That's one approach - another approach would be to say that they're all non-nullable, but nothing checks them. That might be a simpler approach in terms of the spec. I think it's worth discussing this.

Reply:
We discussed this; Nigel will suggest a new name if he thinks of a better one.